File: idea-plugin/src/main/java/com/zzf/rikki/idea/agent/tools/LiteTodoTools.java
package com.zzf.rikki.idea.agent.tools;

import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;

import java.io.File;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;

public class LiteTodoTools {
    private final ObjectMapper mapper;

    public LiteTodoTools(ObjectMapper mapper) {
        this.mapper = mapper;
    }

    public String read(String workspaceRoot, String sessionId) {
        File file = todoFile(workspaceRoot, sessionId);
        if (!file.exists()) {
            return "No todos found.";
        }
        try {
            List<?> todos = mapper.readValue(file, List.class);
            return todos.isEmpty() ? "No todos found." : mapper.writerWithDefaultPrettyPrinter().writeValueAsString(todos);
        } catch (Exception ignored) {
            return "No todos found.";
        }
    }

    public String write(JsonNode args, String workspaceRoot, String sessionId) throws Exception {
        JsonNode todosNode = args.path("todos");
        if (!todosNode.isArray()) {
            throw new IllegalArgumentException("todos must be an array");
        }
        List<Map<String, Object>> todos = new ArrayList<>();
        for (JsonNode item : todosNode) {
            String content = item.path("content").asText("");
            if (content.isBlank()) {
                content = item.path("title").asText("");
            }
            if (content.isBlank()) {
                content = item.path("description").asText("");
            }
            LinkedHashMap<String, Object> todo = new LinkedHashMap<>();
            todo.put("id", item.path("id").asText("").isBlank() ? UUID.randomUUID().toString() : item.path("id").asText(""));
            todo.put("content", content);
            todo.put("status", item.path("status").asText("pending"));
            todo.put("priority", item.path("priority").asText("medium"));
            todos.add(todo);
        }
        File file = todoFile(workspaceRoot, sessionId);
        if (file.getParentFile() != null) {
            file.getParentFile().mkdirs();
        }
        Files.writeString(file.toPath(), mapper.writerWithDefaultPrettyPrinter().writeValueAsString(todos), StandardCharsets.UTF_8);
        return "Todos updated: " + todos.size() + " item(s)";
    }

    public String readJson(String workspaceRoot, String sessionId) {
        File file = todoFile(workspaceRoot, sessionId);
        if (!file.exists()) {
            return null;
        }
        try {
            String text = Files.readString(file.toPath(), StandardCharsets.UTF_8);
            return text.isBlank() ? null : text;
        } catch (Exception ignored) {
            return null;
        }
    }

    private File todoFile(String workspaceRoot, String sessionId) {
        File sessionFile = sessionTodoFile(workspaceRoot, sessionId);
        return sessionFile != null ? sessionFile : new File(workspaceRoot, ".rikki/todos.json");
    }

    private File sessionTodoFile(String workspaceRoot, String sessionId) {
        if (sessionId == null || sessionId.trim().isBlank()) {
            return null;
        }
        return new File(workspaceRoot, ".rikki/todos/" + safeSessionId(sessionId.trim()) + ".json");
    }

    private String safeSessionId(String sessionId) {
        StringBuilder builder = new StringBuilder();
        for (char ch : sessionId.toCharArray()) {
            if (Character.isLetterOrDigit(ch) || ch == '-' || ch == '_' || ch == '.') {
                builder.append(ch);
            } else {
                builder.append('_');
            }
        }
        return builder.toString();
    }
}

File: idea-plugin/src/main/kotlin/com/zzf/rikki/idea/agent/tools/LiteTodoTools.kt
package com.zzf.rikki.idea.agent.tools

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import java.io.File
import java.nio.charset.StandardCharsets
import java.util.UUID

/** Simple file-backed todo storage per workspace + session. */
class LiteTodoTools(private val mapper: ObjectMapper) {

    data class TodoItem(
        val id: String,
        val content: String,
        val status: String,
        val priority: String
    )

    private fun legacyTodoFile(workspaceRoot: String): File =
        File(workspaceRoot, ".rikki/todos.json")

    private fun sessionTodoFile(workspaceRoot: String, sessionId: String): File? {
        val normalized = sessionId.trim()
        if (normalized.isBlank()) return null
        return File(workspaceRoot, ".rikki/todos/${safeSessionId(normalized)}.json")
    }

    private fun todoFile(workspaceRoot: String, sessionId: String): File =
        sessionTodoFile(workspaceRoot, sessionId) ?: legacyTodoFile(workspaceRoot)

    private fun safeSessionId(sessionId: String): String =
        sessionId.map { ch ->
            when {
                ch.isLetterOrDigit() -> ch
                ch == '-' || ch == '_' || ch == '.' -> ch
                else -> '_'
            }
        }.joinToString("")

    fun read(workspaceRoot: String, sessionId: String): String {
        val file = todoFile(workspaceRoot, sessionId)
        if (!file.exists()) return "No todos found."
        return try {
            val todos = mapper.readValue(file, List::class.java)
            if ((todos as List<*>).isEmpty()) "No todos found."
            else mapper.writerWithDefaultPrettyPrinter().writeValueAsString(todos)
        } catch (_: Exception) {
            "No todos found."
        }
    }

    fun write(args: JsonNode, workspaceRoot: String, sessionId: String): String {
        val todosNode = args.path("todos")
        if (!todosNode.isArray) throw IllegalArgumentException("todos must be an array")

        val todos = buildList {
            for (item in todosNode) {
                // Accept 'content' (canonical) or 'title'/'description' as fallbacks
                val content = item.path("content").asText("").ifBlank {
                    item.path("title").asText("").ifBlank { item.path("description").asText("") }
                }
                add(
                    TodoItem(
                        id       = item.path("id").asText("").ifBlank { UUID.randomUUID().toString() },
                        content  = content,
                        status   = item.path("status").asText("pending"),
                        priority = item.path("priority").asText("medium")
                    )
                )
            }
        }

        val file = todoFile(workspaceRoot, sessionId)
        file.parentFile?.mkdirs()
        file.writeText(mapper.writerWithDefaultPrettyPrinter().writeValueAsString(todos), StandardCharsets.UTF_8)
        return "Todos updated: ${todos.size} item(s)"
    }

    fun readJson(workspaceRoot: String, sessionId: String): String? {
        val file = todoFile(workspaceRoot, sessionId)
        if (!file.exists()) return null
        val text = try { file.readText(StandardCharsets.UTF_8) } catch (_: Exception) { return null }
        if (text.isBlank()) return null
        return text
    }
}
